apiserver/internal/handlers/resource/resource.go

```diff
@@ -0,0 +1,417 @@
+		ID:            res.UUID.String(),
+		ApplicationID: res.ApplicationID.String(),
+		Username:      res.RetrievedBy,
+		Timestamp:     res.Timestamp,
+	}
+}
+
+// uploadedResource holds both the information about an uploaded
+// resource and the reader containing its data.
+type uploadedResource struct {
+	// UUID is the resource UUID.
+	UUID coreresource.UUID
+
+	// Application is the Name of the application associated with the resource.
+	Application string
+
+	// Resource is the information about the resource.
+	Resource charmresource.Resource
+
+	// Size is the size of the resource blob.
+	Size int64
+
+	// Fingerprint is the hash of the resource blob.
+	Fingerprint charmresource.Fingerprint
+}
+
+// readResource extracts the relevant info from the request.
+func (h *ResourceHandler) readResource(resourceService ResourceService, req *http.Request) (io.ReadCloser, *uploadedResource, error) {
+	defer func() {
+		if req.Body != nil {
+			req.Body.Close()
+		}
+	}()
+	uReq, err := extractUploadRequest(req)
+	if err != nil {
+		return nil, nil, errors.Capture(err)
+	}
+
+	uuid, err := resourceService.GetResourceUUIDByApplicationAndResourceName(req.Context(), uReq.Application, uReq.Name)
+	if errors.Is(err, resourceerrors.ResourceNotFound) {
+		return nil, nil, jujuerrors.NotFoundf("resource %s of application %s", uReq.Name, uReq.Application)
+	} else if err != nil {
+		return nil, nil, errors.Errorf("getting resource uuid: %w", err)
+	}
+
+	res, err := resourceService.GetResource(req.Context(), uuid)
+	if err != nil {
+		return nil, nil, errors.Errorf("getting resource details: %w", err)
+	}
+
+	switch res.Type {
+	case charmresource.TypeFile:
+		ext := path.Ext(res.Path)
+		if path.Ext(uReq.Filename) != ext {
+			return nil, nil, errors.Errorf("incorrect extension on resource upload %q, expected %q", uReq.Filename, ext)
+		}
+	}
+
+	reader, err := h.validator.Validate(req.Body, res.Fingerprint.String(), res.Size)
+	if err != nil {
+		return nil, nil, errors.Errorf("validating resource size and hash: %w", err)
+	}
+
+	return reader, &uploadedResource{
+		UUID:        res.UUID,
+		Application: uReq.Application,
+		Resource:    res.Resource,
+		Size:        uReq.Size,
+		Fingerprint: uReq.Fingerprint,
+	}, nil
+}
+
+// extractUploadRequest pulls the required info from the HTTP request.
+func extractUploadRequest(req *http.Request) (api.UploadRequest, error) {
+	var ur api.UploadRequest
+
+	if req.Header.Get(api.HeaderContentLength) == "" {
+		size := req.ContentLength
+		// size will be negative if there is no content.
+		if size < 0 {
+			size = 0
+		}
+		req.Header.Set(api.HeaderContentLength, fmt.Sprint(size))
+	}
+
+	ctype := req.Header.Get(api.HeaderContentType)
+	if ctype != api.ContentTypeRaw {
+		return ur, errors.Errorf("unsupported content type %q", ctype)
+	}
+
+	application, name := extractEndpointDetails(req.URL)
+	fingerprint := req.Header.Get(api.HeaderContentSha384) // This parallels "Content-MD5".
```

Comment:
What do you mean by the comment?

Reply:
This was there before. Its no longer clear where the Content-MD5 is that its paralleling, I've removed the comment.